### PR TITLE
First refactoring (thaw_fast)

### PR DESCRIPTION
**Pull request by fparain**
Refactoring method thaw_fast() for better encapsulation and readability.

---------
### Progress
- [x] Change must not contain extraneous whitespace

### Reviewers
 * [Coleen Phillimore](https://openjdk.java.net/census#coleenp) (@coleenp - Committer) ⚠️ Review applies to 5470e42e23689f7926ea5667d2414c8b48c8e4a9
 * [Robbin Ehn](https://openjdk.java.net/census#rehn) (@robehn - Committer) ⚠️ Review applies to 5470e42e23689f7926ea5667d2414c8b48c8e4a9

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/loom pull/158/head:pull/158` \
`$ git checkout pull/158`

Update a local copy of the PR: \
`$ git checkout pull/158` \
`$ git pull https://git.openjdk.java.net/loom pull/158/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 158`

View PR using the GUI difftool: \
`$ git pr show -t 158`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/loom/pull/158.diff">https://git.openjdk.java.net/loom/pull/158.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp [hunk 1]
```diff
@@ -1626,6 +1626,48 @@ inline intptr_t* Thaw<ConfigT>::thaw(thaw_kind kind) {
...
+    assert(argsize != 0 || (_base - _thaw_size) == ContinuationHelper::frame_align_pointer(_base - _thaw_size), "");
+  }
+  int thaw_size() const { return _thaw_size; }
+  int argsize() const { return _argsize; }
+
+  // top and bottom stack pointers
+  intptr_t* sp() const { return ContinuationHelper::frame_align_pointer(_base - _thaw_size); }
+  intptr_t* bottom_sp() const { return ContinuationHelper::frame_align_pointer(_base - _argsize); }
+
+  // several operations operate ont the totality of the stack being reconstructed,
```
Comment (coleenp) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:1647]: typo

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp [hunk 2]
```diff
@@ -1626,6 +1626,48 @@ inline intptr_t* Thaw<ConfigT>::thaw(thaw_kind kind) {
                                         : thaw_slow(chunk, kind != thaw_top);
 }
 
+class ReconstructedStack {
```
Comment (coleenp) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:1629]: pls inherit from StackObj

**Review by coleenp — COMMENTED** [review]
Yes, this is nice and helpful.

File: src/hotspot/share/oops/stackChunkOop.inline.hpp [hunk 3]
```diff
@@ -167,6 +167,39 @@ inline bool stackChunkOopDesc::requires_barriers() {
...
+inline void stackChunkOopDesc::clear_chunk() {
+  set_sp(stack_size());
+  set_argsize(0);
+  set_max_thawing_size(0);
+}
+
+inline int stackChunkOopDesc::remove_top_compiled_frame(int &argsize) {
+  bool empty = false;
+  StackChunkFrameStream<ChunkFrames::CompiledOnly> f(this);
+  intptr_t* const chunk_sp = start_address() + sp();
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/oops/stackChunkOop.inline.hpp:179]: Yes, chunk_sp is the derelativized value of sp().
chunk_sp is only used in asserts. Some asserts are specific to the current case of a compiled method.

File: src/hotspot/share/oops/stackChunkOop.inline.hpp [hunk 4]
```diff
@@ -167,6 +167,39 @@ inline bool stackChunkOopDesc::requires_barriers() {
...
+  } else {
+    set_sp(sp() + frame_size);
+    set_max_thawing_size(max_thawing_size() - frame_size);
+    // We set chunk->pc to the return pc into the next frame
+    set_pc(f.pc());
+    assert(f.pc() == *(address*)(chunk_sp + frame_size - frame::sender_sp_ret_address_offset()), "unexpected pc");
+  }
+  assert(empty == is_empty(), "");
+  return frame_size + argsize;
+}
```
[the remark this answers is not in the document]
Comment (fparain) [reply src/hotspot/share/oops/stackChunkOop.inline.hpp:203]: Comment added

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp (hunk 1 above)
Comment (fparain) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp]: Fixed

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp (hunk 2 above)
Comment (fparain) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp]: Inheritance added.

**Review by robehn — APPROVED** [review]
I think overall it's an improvement.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp [hunk 5]
```diff
@@ -1715,15 +1717,50 @@ NOINLINE intptr_t* Thaw<ConfigT>::thaw_fast(stackChunkOop chunk) {
...
+    chunk->set_sp(chunk->sp() + frame_size);
+    chunk->set_max_thawing_size(chunk->max_thawing_size() - frame_size);
+    // We set chunk->pc to the return pc into the next frame
+    chunk->set_pc(f.pc());
+    assert(f.pc() == *(address*)(chunk_sp + frame_size - frame::sender_sp_ret_address_offset()), "unexpected pc");
+  }
+  assert(empty == chunk->is_empty(), "");
+  // returns the size required to store the frame on stack, and because it is a
+  // compiled frame, it must include a copy of the arguments passed by the caller
+  return frame_size + argsize;
```
Comment (coleenp) [inline src/hotspot/share/runtime/continuationFreezeThaw.cpp:1763]: small nit, can you put these functions above the ones that call them in the file?

**Review by coleenp — APPROVED** [review]
Looks good.

File: src/hotspot/share/runtime/continuationFreezeThaw.cpp (hunk 5 above)
Comment (fparain) [reply src/hotspot/share/runtime/continuationFreezeThaw.cpp]: Sure. Methods re-ordered: clear_chunk(),  remove_top_compiled_frame_from_chunk(), copy_from_chunk() and patch_return().

**Pull request closed without merging**